### PR TITLE
[WIP] Optimize calls to `copy_nonoverlapping` intrinsic to assignment 

### PR DESCRIPTION
**Pull request by wesleywiser**
r? @ghost

Rebased on top of #81238

### CODE REVIEW

**Comment by bugadani** [comment]
Fixes #73258 and I wonder how many more :) Nice!

File: compiler/rustc_mir/src/transform/instcombine.rs [hunk 1]
```diff
@@ -120,4 +127,78 @@ impl<'tcx, 'a> InstCombineContext<'tcx, 'a> {
...
+        None
+    }
+
+    fn combine_copy_nonoverlapping(
+        &self,
+        terminator: &mut Terminator<'tcx>,
+        locals: &LocalDecls<'tcx>,
+        statements: &mut Vec<Statement<'tcx>>,
+    ) {
+        if let Some((src, dest, next_bb)) = self.find_copy_nonoverlapping(terminator, locals) {
```
Comment (wesleywiser) [inline compiler/rustc_mir/src/transform/instcombine.rs:189]: TODO: respect optimization fuel

**Comment by bjorn3** [comment]
Many improvements of up to 2.6% with a few small regressions of up to 0.8%.

File: compiler/rustc_mir/src/transform/instcombine.rs [hunk 2]
```diff
@@ -25,6 +28,10 @@ impl<'tcx> MirPass<'tcx> for InstCombine {
                     _ => {}
                 }
             }
+
+            if let Some(terminator) = &mut block.terminator {
```
Comment (bjorn3) [inline compiler/rustc_mir/src/transform/instcombine.rs:32]: You should use `block.terminator_mut()`. The terminator is always `Some` after mir building.

**Comment by nagisa** [comment]
Shouldn't this be a `lower_intrinsics` thing, rather than a `instcombine`?

**Comment by nagisa** [comment]
cc https://github.com/rust-lang/compiler-team/issues/348 and https://github.com/rust-lang/rust/pull/77511.

**Comment by wesleywiser** [comment]
Oh, I'm sorry! I'd completely forgotten about that MCP. 

Closing in favor of #77511

**Pull request closed without merging**